File: src/test/java/io/primeaspect/calculator/service/StatisticsServiceTest.java
package io.primeaspect.calculator.service;

import io.primeaspect.calculator.dto.response.ExpressionListResponse;
import io.primeaspect.calculator.model.Expression;
import io.primeaspect.calculator.model.Number;
import io.primeaspect.calculator.repository.ExpressionRepository;
import io.primeaspect.calculator.repository.NumberStatisticsRepository;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;
import org.mockito.Mockito;

import java.time.LocalDate;
import java.util.ArrayList;
import java.util.List;

import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

public class StatisticsServiceTest {
    private final ExpressionRepository expressionRepository = Mockito.mock(ExpressionRepository.class);
    private final NumberStatisticsRepository numberRepository = Mockito.mock(NumberStatisticsRepository.class);
    private final StatisticsService service = new StatisticsService(expressionRepository, numberRepository);

    @Test
    public void findAllByDateTest() {
        LocalDate date = LocalDate.of(2017, 12, 15);
        Expression expression = new Expression(date, "(-78+9-(9/4.5))^2", 5401.0);
        List<Expression> expectedList = new ArrayList<>();
        expectedList.add(expression);
        ExpressionListResponse expected = new ExpressionListResponse(expectedList);

        when(expressionRepository.findAllByDate(date)).thenReturn(expectedList);

        ExpressionListResponse response = service.findAllByDate(date.toString());

        verify(expressionRepository).findAllByDate(date);
        Assertions.assertEquals(expected, response);
    }

    @Test
    public void findAllByOperationTest() {
        LocalDate date = LocalDate.of(2017, 12, 15);
        String request = "^";
        Expression expression = new Expression(date, "(-78+9-(9/4.5))^2", 5401.0);
        List<Expression> expectedList = new ArrayList<>();
        expectedList.add(expression);
        ExpressionListResponse expected = new ExpressionListResponse(expectedList);

        when(expressionRepository.findAllByOperation(request)).thenReturn(expectedList);

        ExpressionListResponse response = service.findAllByOperation(request);

        verify(expressionRepository).findAllByOperation(request);
        Assertions.assertEquals(expected, response);
    }

    @Test
    public void getCountByDateTest() {
        LocalDate date = LocalDate.of(2017, 12, 15);
        when(expressionRepository.getCountByDate(date)).thenReturn(5);

        int response = service.getCountByDate(date.toString());

        verify(expressionRepository).getCountByDate(date);
        Assertions.assertEquals(5, response);
    }

    @Test
    public void getCountByOperationTest() {
        String request = "^";
        when(expressionRepository.getCountByOperation(request)).thenReturn(5);

        int response = service.getCountByOperation(request);

        verify(expressionRepository).getCountByOperation(request);
        Assertions.assertEquals(5, response);
    }

    @Test
    public void getPopularNumberTest() {
        Number expected = new Number(10, 5);

        when(numberRepository.getPopularNumber()).thenReturn(expected);

        Number response = service.getPopularNumber();

        verify(numberRepository).getPopularNumber();
        Assertions.assertEquals(expected, response);
    }
}

File: src/test/kotlin/io/primeaspect/calculator/service/StatisticsServiceTest.kt
package io.primeaspect.calculator.service

import com.nhaarman.mockitokotlin2.mock
import com.nhaarman.mockitokotlin2.verify
import com.nhaarman.mockitokotlin2.whenever
import io.primeaspect.calculator.dto.response.ExpressionListResponse
import io.primeaspect.calculator.dto.response.NumberResponse
import io.primeaspect.calculator.model.Expression
import io.primeaspect.calculator.repository.ExpressionRepository
import io.primeaspect.calculator.repository.impl.mybatis.NumberStatisticsRepositoryMyBatis
import org.junit.jupiter.api.Test
import java.time.LocalDate
import kotlin.test.assertEquals

class StatisticsServiceTest {
    private val expressionRepository: ExpressionRepository = mock()
    private val numberRepository: NumberStatisticsRepositoryMyBatis = mock()
    private val service = StatisticsService(expressionRepository, numberRepository)

    @Test
    fun `The findAllByDate(date) method of the expressionRepository is called when the findAllByDate(date) method of the service is called`() {
        val date = LocalDate.of(2017, 12, 15)
        val expression = Expression(date = date, expression = "(-78+9-(9/4.5))^2", solution = 5401.0)
        val expectedList: List<Expression> = listOf(expression)
        val (expected) = ExpressionListResponse(expectedList)

        whenever<List<Expression>>(expressionRepository.findAllByDate(date)).thenReturn(expectedList)

        val (response) = service.findAllByDate(date)

        verify(expressionRepository).findAllByDate(date)
        assertEquals(expected, response)
    }

    @Test
    fun `The findAllByOperation(request) method of the expressionRepository is called when the findAllByOperation(request) method of the service is called`() {
        val date = LocalDate.of(2017, 12, 15)
        val request = "^"
        val expression = Expression(date = date, expression = "(-78+9-(9/4.5))^2", solution = 5401.0)
        val expectedList: List<Expression> = listOf(expression)
        val (expected) = ExpressionListResponse(expectedList)

        whenever(expressionRepository.findAllByOperation(request)).thenReturn(expectedList)

        val (response) = service.findAllByOperation(request)

        verify(expressionRepository).findAllByOperation(request)
        assertEquals(expected, response)
    }

    @Test
    fun `The getCountByDate(date) method of the expressionRepository is called when the getCountByDate(date) method of the service is called`() {
        val date = LocalDate.of(2017, 12, 15)
        whenever(expressionRepository.getCountByDate(date)).thenReturn(5)

        val response = service.getCountByDate(date)

        verify(expressionRepository).getCountByDate(date)
        assertEquals(5, response)
    }

    @Test
    fun `The getCountByOperation(request) method of the expressionRepository is called when the getCountByOperation(request) method of the service is called`() {
        val request = "^"
        whenever(expressionRepository.getCountByOperation(request)).thenReturn(5)

        val response = service.getCountByOperation(request)

        verify(expressionRepository).getCountByOperation(request)
        assertEquals(5, response)
    }

    @Test
    fun `The getPopularNumber() method of the numberRepository is called when the getPopularNumber() method of the service is called`() {
        val expected = NumberResponse(value = 10.0, count = 5)

        whenever(numberRepository.getPopularNumber()).thenReturn(expected)

        val response = service.getPopularNumber()

        verify(numberRepository).getPopularNumber()
        assertEquals(expected.value, response.value)
        assertEquals(expected.count, response.count)
    }

    @Test
    fun `The getPopularNumberByDate() method of the numberRepository is called when the getPopularNumberByDate() method of the service is called`() {
        val date = LocalDate.of(2017, 12, 15)
        val expected = NumberResponse(value = 10.0, count = 5)

        whenever(numberRepository.getPopularNumberByDate(date)).thenReturn(expected)

        val response = service.getPopularNumberByDate(date)

        verify(numberRepository).getPopularNumberByDate(date)
        assertEquals(expected.value, response.value)
        assertEquals(expected.count, response.count)
    }
}